### PR TITLE
Encoder direction: Wrangle, Revert select screen change

### PR DESCRIPTION
**Pull request by robbycandra**
I use Reprap Smart Discount LCD 2004

This commit [https://github.com/marlinfirmware/marlin/commit/3cd9a92dcc525e97768330651c40e5dc349c9e19](https://github.com/marlinfirmware/marlin/commit/3cd9a92dcc525e97768330651c40e5dc349c9e19
), makes me cannot reverse Encoder Direction in do_select_screen.
It working good before this commit.

REVERSE_ENCODER_DIRECTION already have effect in do_select_screen.

```
ultralcd.h 


    #if ENABLED(REVERSE_ENCODER_DIRECTION)
      #define ENCODERBASE -1
    #else
      #define ENCODERBASE +1
    #endif
    #if ENABLED(REVERSE_MENU_DIRECTION)
      static int8_t encoderDirection;
      static inline void encoder_direction_normal() { encoderDirection = +(ENCODERBASE); }
      static inline void encoder_direction_menus()  { encoderDirection = -(ENCODERBASE); }
    #else
      static constexpr int8_t encoderDirection = ENCODERBASE;
      static inline void encoder_direction_normal() {}
      static inline void encoder_direction_menus()  {}
    #endif
```

```
ultralcd.cpp
      // Manage encoder rotation
      #define ENCODER_SPIN(_E1, _E2) 
        switch (lastEncoderBits) { 
          case _E1: encoderDiff += encoderDirection; 
                        break; 
          case _E2: encoderDiff -= encoderDirection; 
        }
...
          encoderPosition += (encoderDiff * encoderMultiplier) / (ENCODER_PULSES_PER_STEP);
          encoderDiff = 0;

```

### CODE REVIEW

**Comment by thinkyhead** [comment]
This PR fails to preserve the intent of the change. The logic is apparently off, but what we want is for the "value increase" direction to select the RIGHT option and for the "value decrease" direction to select the LEFT option.

File: Marlin/src/lcd/menu/menu.cpp [hunk 1]
```diff
@@ -448,7 +450,7 @@ bool ui_selection; // = false
 void set_ui_selection(const bool sel) { ui_selection = sel; }
 void do_select_screen(PGM_P const yes, PGM_P const no, selectFunc_t yesFunc, selectFunc_t noFunc, PGM_P const pref, const char * const string/*=nullptr*/, PGM_P const suff/*=nullptr*/) {
   if (ui.encoderPosition) {
-    ui_selection = ((ENCODERBASE) > 0) == (int16_t(ui.encoderPosition) > 0);
+    ui_selection = ((encoderDirection > 0) == int16_t(ui.encoderPosition) > 0);
```
Comment (thinkyhead) [inline Marlin/src/lcd/menu/menu.cpp]: As an experiment, try the commits I've added to see if it's better, and also add this here:
```cpp
  SERIAL_ECHOLNPAIR("ENCODERBASE=", int(ENCODERBASE),
                    " encoderDirection=", int(encoderDirection),
                    " encPos>0=`, int(ui.encoderPosition > 0)
  );
```
to see the values in action. If it's still broken, then attach your configurations to the reply so we can see why wrong values are coming out. From all these pieces we should be able to get it sorted quickly.

Comment (robbycandra) [reply Marlin/src/lcd/menu/menu.cpp]: My point is 
```
ui_selection = ((ENCODERBASE) > 0) == (int16_t(ui.encoderPosition) > 0);
```
This make the ui.selection REVERSED twice.

First  in ultralcd.h
```
 static constexpr int8_t encoderDirection = ENCODERBASE;
```

Second in the reverted commit.

All is good before. And yes, i dont know the intent of the reverted commit. But i will try.

Comment (thinkyhead) [reply Marlin/src/lcd/menu/menu.cpp]: I forget who it was that needed the change, but it worked to fix the problem in their setup. Perhaps it needs refinement for keypads and other non-wheel setups too…?

Comment (thinkyhead) [reply Marlin/src/lcd/menu/menu.cpp]: > This make the ui.selection REVERSED twice.

Presumably then, replacing `ENCODERBASE` with `encoderDirection` should work. When in the selection screen, the direction should be as if inside an edit (or any non-menu) screen. I hope this tweak will do the trick.

**Comment by robbycandra** [comment]
Why not use 
```
//
// This option reverses the encoder direction for navigating LCD menus.
//
//  If CLOCKWISE normally moves DOWN this makes it go UP.
//  If CLOCKWISE normally moves UP this makes it go DOWN.
//
//#define REVERSE_SELECT_SCREEN_DIRECTION
```
Its simple. 
Just like REVERSE_MENU_DIRECTION.

**Comment by marcio-ao** [comment]
This may actually be closely related to #14659 which I just created. I don't think there needs to be a separate option reversing the encoder direction in the start print dialog box, it merely needs to be consistent with the menus.

**Comment by marcio-ao** [comment]
Not sure whether this PR or mine is the most correct, but mine is a lot shorter.

@robbycandra: can you look at my PR and see whether it addresses the problem as well as yours? Do you see a reason for one vs another?

**Comment by robbycandra** [comment]
@marcio-ao in my opinion, It only need to be reverted. Your PR is not correct too because if "REVERSE_ENCODER_DIRECTION enabled or not, it will not change encoder direction in do select screen.

**Comment by thinkyhead** [comment]
> Not sure whether this PR or mine is the most correct, but mine is a lot shorter.

I tacked on corrections for the encoder direction state so the direction is always "normal" on entry to a new screen, and always set to "menu" when that screen is a menu (or static scrolling display). I'll check out your change in a sec and see what it's doing….

**Comment by thinkyhead** [comment]
I will be curious to know whether this works, since #14659 is the same as ee1a84b.

**Comment by thinkyhead** [comment]
> Just like REVERSE_MENU_DIRECTION.

The rotation direction that makes the FR percentage increase on the main screen
should also cause the RIGHT option to be selected on a selection screen.

**Comment by robbycandra** [comment]
@marcio-ao , its already merged. 

Btw... i am learing about Extensible UI now... It is amazing concept.

**Comment by marcio-ao** [comment]
> The rotation direction that makes the FR percentage increase on the main screen
should also cause the RIGHT option to be selected on a selection screen.

@thinkyhead: I don't agree with this statement. If you look at the comments, you see there are two separate actions. One, increments numerical values, and the second increments menu choices:

 *  Reversed Value Edit and Menu Nav? Enable REVERSE_ENCODER_DIRECTION.
 *  Reversed Menu Navigation only?    Enable REVERSE_MENU_DIRECTION.
 *  Reversed Value Editing only?      Enable BOTH options.

There is no third category for the selection of choices in the a dialog box, but I think it should mirror menus, because effectively it is a menu of two choices, just laid out horizontally.

So, in short, all numerical editors, including feedrate, need to respect REVERSE_ENCODER_DIRECTION and ignore REVERSE_MENU_DIRECTION.  But menu item and selection of choices in a dialog box need to respect both REVERSE_ENCODER_DIRECTION and REVERSE_MENU_DIRECTION.

Note that this means a user could configure the selection of item choices to happen clockwise, while having numerical values (including feedrates) go counter-clockwise.

**Comment by marcio-ao** [comment]
So I just checked the latest upstream code, as of commit 1ea9a36d, and its all FUBAR now. Here is a table of the effect of REVERSE_ENCODER_DIRECTION and REVERSE_MENU_DIRECTION:

| RED | RMD | FR+ | Temp+ | Menu Down | Sel Right | 
|-----|-----|-----|-----|-----|-----|
| false | false | ~CW~ | CCW | CCW | CCW | ~CW~ | CCW |
| false | true |  ~CW~ | ~CW~ | CW  |  ~CCW~ | 
| true  | false | CW | CW | CW | CW | 
| true | true |  CW | ~CCW~ | CCW | ~CW~ | 

All the directions indicated by the strikethrough are incorrect. The table should look like this:

| RED | RMD | FR+ | Temp+ | Menu Down | Sel Right | 
|-----|-----|-----|-----|-----|-----|
| false | false | CCW | CCW | CCW | CCW | 
| false | true |  CCW | CCW | CW  |  CW | 
| true  | false | CW | CW | CW | CW | 
| true | true | CW | CW | CCW | CCW | 

To be honest, things were a lot better before this was PR was merged in. Now it's just a complete mess.

**Comment by marcio-ao** [comment]
Notice that the table is fairy simple FR+ and Temp+ should both be coordinated, as they are both numeric values. Hence, the first two columns only depend on REVERSE_ENCODER_DIRECTION (RED). Setting REVERSE_MENU_DIRECTION (RMD) should flip the state of the Menu Down and Sel Right columns. Here's a way to represent it in boolean logic:

| FR+ | Temp+ | Menu Down | Sel Right |
|-----|-----|------|------|
| RED | RED | RED ^ RMD | RED ^ RMD |

Easy peasy!

**Comment by robbycandra** [comment]
@marcio-ao 
it is very clear if we look at ultralcd.h

```
#if ENABLED(REVERSE_ENCODER_DIRECTION)
      #define ENCODERBASE -1
    #else
      #define ENCODERBASE +1
    #endif
    #if ENABLED(REVERSE_MENU_DIRECTION)
      static int8_t encoderDirection;
      static inline void encoder_direction_normal() { encoderDirection = +(ENCODERBASE); }
      static inline void encoder_direction_menus()  { encoderDirection = -(ENCODERBASE); }
    #else
      static constexpr int8_t encoderDirection = ENCODERBASE;
      static inline void encoder_direction_normal() {}
      static inline void encoder_direction_menus()  {}
    #endif
```

**Comment by robbycandra** [comment]
For me : RED = True and RMD = False are good combination.

**Comment by marcio-ao** [comment]
@robbycandra: You are exactly right. RED = True and RMD = False is the *only* combination that is correct in the table I presented above. All the other combinations lead to non-sensical garbage.

We need a strategy for tacking this problem. First, we need to all agree that the table I proposed is what we are aiming for. Then, we need to come up with a fix (or fixes) that works for all combinations of RED and RMD, not just one combination.

**Comment by marcio-ao** [comment]
So lets start with the code you posted, why doesn't it look like this?

```
#if ENABLED(REVERSE_ENCODER_DIRECTION)
      #define ENCODERBASE -1
    #else
      #define ENCODERBASE +1
    #endif
    #if ENABLED(REVERSE_MENU_DIRECTION)
      static int8_t encoderDirection;
      static inline void encoder_direction_normal() { encoderDirection = +(ENCODERBASE); }
      static inline void encoder_direction_menus()  { encoderDirection = -(ENCODERBASE); }
    #else
      static constexpr int8_t encoderDirection = ENCODERBASE;
      static inline void encoder_direction_normal() {encoderDirection = -(ENCODERBASE); }
      static inline void encoder_direction_menus()  {encoderDirection = +(ENCODERBASE); }
    #endif
```

**Comment by marcio-ao** [comment]
@robbycandra: I guess I was kind of mean in saying it was this PR that caused the entire mess. I apologize. I do not know for a fact that it was this PR that did it. But something changed significantly since the last pull I made and now the LCD encoder directions are all over the place.

**Comment by marcio-ao** [comment]
Okay, I think I am starting to understand this. The other half of the equation is the lines [here](https://github.com/MarlinFirmware/Marlin/blob/1ea9a36d11cdf888637455df6a11e64dd62478a1/Marlin/src/lcd/ultralcd.cpp#L184):

  ```
#if ENABLED(REVERSE_MENU_DIRECTION)
    int8_t MarlinUI::encoderDirection = 1;
  #14182 
```

I wonder if it would make a difference to initialize that to ENCODERBASE rather than 1.

**Comment by robbycandra** [comment]
@marcio-ao:
I haven't try the result after this PR is merged. But my intention is only to revert the change in do select screen function and return to the previous code before this encoder problem appear.  Please take a look at commit history above

Encoder

I Make this table, before this PR merged. But when i want to post it, the PR already merged and now i havent test it.

**Comment by robbycandra** [comment]
This encoder code become too complicated, Actually i prefer to revert all and back to old code.

**Comment by robbycandra** [comment]
image

I agree, it same with this:

image

**Comment by marcio-ao** [comment]
@robbycandra: I'm going to study the code more and see if I can make heads and tails out of this.

One area where the current implementation seems problematic is that when `REVERSE_MENU_DIRECTION` is enabled, the encoder direction is a run-time state variable that needs to be initialized by a call to either `encoder_direction_normal()` or `encoder_direction_menu()`. It seems like it would be easy to forget this call, in which case the direction of the encoder would depend on the menu you left, rather than the current menu you were at. To me, this seems error prone.

It seems to me like REVERSE_MENU_DIRECTION could be eliminated altogether and greatly simplify the code. IMHO, the direction for increasing values should be the same as the direction of going down the menu, or selecting the next choice in a dialog box -- as all of these connote moving "forward" to what is next. Whether that direction is CCW or CW could be controlled by REVERSE_ENCODER_DIRECTION, but I really don't see a need to reverse the direction of values vs menus.

**Comment by marcio-ao** [comment]
Okay, I think I got a handle on this:

- The `encoder_direction_menus()` is only called in one place. It is called in `scroll_screen` [here](https://github.com/MarlinFirmware/Marlin/blob/1ea9a36d11cdf888637455df6a11e64dd62478a1/Marlin/src/lcd/menu/menu.cpp#L328). So, this means that any screen that scrolls (i.e. menus) will be affected by both REVERSE_MENU_DIRECTION.
- However, the dialog boxes that allow the selection of choices are drawn by `do_select_screen`. In order for these to behave like menus, it will be necessary to call `encoder_direction_menus()` [here](https://github.com/MarlinFirmware/Marlin/blob/1ea9a36d11cdf888637455df6a11e64dd62478a1/Marlin/src/lcd/menu/menu.cpp#L462) too.

Once that change is done, the the menu and the choice dialog boxes will behave the same, fixing the two last columns in the table:

| RED | RMD | FR+ | Temp+ | Menu Down | Sel Right | 
|-----|-----|-----|-----|-----|-----|
| false | false | ~CW~ | CCW | CCW | CCW | ~CW~ | CCW |
| false | true |  ~CW~ | ~CW~ | CW  | CW | 
| true  | false | CW | CW | CW | CW | 
| true | true |  CW | ~CCW~ | CCW | CCW | 

Now, the temperature screen is handled by `draw_edit_screen`. Since we do not want this to be affected by REVERSE_MENU_DIRECTION, I think it needs to call `encoder_direction_normal()`. However, I do not see that call anywhere, so it probably takes the direction of the last screen, which happens to be a menu.

Last but not least, FR+ is part of the status screen. It currently is invariant to either REVERSE_MENU_DIRECTION and REVERSE_ENCODER_DIRECTION, so it must be implemented differently than the others. We need to find the code and fix it so it operates with `encoder_direction_normal()`

Anyhow, at this point I haven't tested any of these theories, but I feel I am close.

**Comment by marcio-ao** [comment]
@robbycandra: See PR #14673, I believe I got everything back to where it needed to be with a three line change!

**Comment by AnHardt** [comment]
How about organizing left/right as a ring?
With only two items any direction turn would just select the other item.

**Comment by marcio-ao** [comment]
> How about organizing left/right as a ring?
> With only two items any direction turn would just select the other item.

~Nah. This would require the user to turn the knob carefully to select the option they wanted without overshooting. As it is, with a hard stop, the user can just do a quick flick of the wrist to change the selection. I like it.~ **Edit:** Nevermind, I get it now.

**Comment by marcio-ao** [comment]
@AnHardt: Oh, I see what you mean, a flick right would move to the other item, but stay put on it even if you continued turning; likewise, going to the left would select the right menu item, but then stay there even if you continued turning. That is a possible alternative.

**Comment by marcio-ao** [comment]
@AnHardt: My current PR (#14673) makes it so the dialog box requires a turn in the same direction as going down the menu. For me, this is sufficient and intuitive. But if you want to make an additional PR that allows a selection in both directions, that certainly is fine too.

**Comment by robbycandra** [comment]
@AnHardt , maybe in the future we have 3 option for LEFT / RIGHT menu. 
And without ring, we can turn encoder faster because we can turn Left or Right without thinking how many click.
